### PR TITLE
Remove some duplicates and run typo3-rector

### LINKED ISSUE #63: Output of location not consistent
To output the location in an email or .ics file a $helper array is used. Ignoring that the same code is used in different places (I'll do a refactoring) I want to focus on these two code blocks.

SubscriberController.php:234
```
if (is_object($event->getLocation())) {
    if (is_object($event->getLocation()->getParent()->current())) {
        $helper['location'] = $event->getLocation()->getParent()->current()->getName() . ', ';
        $helper['locationics'] = $this->foldline($event->getLocation()->getParent()->current()->getName()) . ', ';
    }
    $helper['location'] .= $event->getLocation()->getName();
    $helper['locationics'] .= $this->foldline($event->getLocation()->getName());
}
```

and SubscriberController.php:611
```
if (is_object($event->getLocation())) {
    if (is_object($event->getLocation()->getParent()->current())) {
        $helper['location'] = $event->getLocation()->getParent()->current()->getName() . ', ';
        $helper['locationics'] =
            $this->foldline($event->getLocation()->getParent()->current()->getName()) . ', ';
    }
    $helper['location'] = $event->getLocation()->getName();
    $helper['locationics'] = $this->foldline($event->getLocation()->getName());
}
```

The only difference is the missing string concatenation in the second block. It seems that a concatenation was intended, because of the comma three lines earlier, but the complete gets overridden by the child.

Whats' the wanted behavior?

If the concatenation is wanted another problem comes up, because in the first code block two folded lines were put together.

IMHO: The content composition for locations and their parents should happen in the template, also the foldline-method should be a ViewHelper. 

I can do this, if you give me you opinion. Maybe it also makes sense to put $helper into a separate class or remove $helper and put some logic into the Event::class

### PR DESCRIPTION
**Pull request by PeaceSfiSh**
I tried to remove some duplicated code blocks and fixed #63.

typo3-rector configuration

```
<?php

declare(strict_types=1);

use Rector\Core\Configuration\Option;
use Rector\Core\ValueObject\PhpVersion;
use Rector\PostRector\Rector\NameImportingPostRector;
use Symfony\Component\DependencyInjection\Loader\Configurator\ContainerConfigurator;
use Ssch\TYPO3Rector\Set\Typo3SetList;
use Ssch\TYPO3Rector\Configuration\Typo3Option;
use Ssch\TYPO3Rector\Rector\v9\v0\InjectAnnotationRector;

return static function (ContainerConfigurator $containerConfigurator): void {
    // get parameters
    $parameters = $containerConfigurator->parameters();

    // Define what rule sets will be applied
    $parameters->set(Option::SETS, [
        Typo3SetList::TYPO3_76,
        Typo3SetList::TYPO3_87,
        Typo3SetList::TYPO3_95
    ]);

    // FQN classes are not imported by default. If you don't do it manually after every Rector run, enable it by:
    $parameters->set(Option::AUTO_IMPORT_NAMES, true);

    // this will not import root namespace classes, like \DateTime or \Exception
    $parameters->set(Option::IMPORT_SHORT_CLASSES, false);

    // this will not import classes used in PHP DocBlocks, like in /** @var \Some\Class */
    $parameters->set(Option::IMPORT_DOC_BLOCKS, false);

    // Define your target version which you want to support
    $parameters->set(Option::PHP_VERSION_FEATURES, PhpVersion::PHP_72);

    // If you would like to see the changelog url when a rector is applied
    $parameters->set(Typo3Option::OUTPUT_CHANGELOG, true);

    // If you set option Option::AUTO_IMPORT_NAMES to true, you should consider excluding some TYPO3 files.
    $parameters->set(Option::SKIP, [
        NameImportingPostRector::class => [
            'ClassAliasMap.php',
            'ext_localconf.php',
            'ext_emconf.php',
            'ext_tables.php',
            __DIR__ . '/**/TCA/*',
            __DIR__ . '/**/Tests/*',
            __DIR__ . '/**/Configuration/RequestMiddlewares.php',
            __DIR__ . '/**/Configuration/Commands.php',
            __DIR__ . '/**/Configuration/AjaxRoutes.php',
            __DIR__ . '/**/Configuration/Extbase/Persistence/Classes.php',
        ],
        __DIR__ . '/**/Tests/*'
    ]);

    // If you have trouble that rector cannot run because some TYPO3 constants are not defined add an additional constants file
    // Have a look at https://github.com/sabbelasichon/typo3-rector/blob/master/typo3.constants.php
    // $parameters->set(Option::AUTOLOAD_PATHS, [
    //    __DIR__ . '/typo3.constants.php'
    // ]);

    // get services (needed for register a single rule)
    // $services = $containerConfigurator->services();

    // register a single rule
    // $services->set(InjectAnnotationRector::class);
};
```

### CODE REVIEW

File: Classes/Controller/LocationController.php [hunk 1]
```diff
@@ -1,6 +1,7 @@
 <?php
 namespace Slub\SlubEvents\Controller;
 
+use TYPO3\CMS\Extbase\Mvc\Controller\ActionController;
```
Comment (albig) [inline Classes/Controller/LocationController.php:4]: Please put the use statement below the copyright notice.

File: Classes/Controller/SubscriberController.php [hunk 2]
```diff
@@ -1,6 +1,7 @@
 <?php
 namespace Slub\SlubEvents\Controller;
 
+use TYPO3\CMS\Core\Page\PageRenderer;
```
Comment (albig) [inline Classes/Controller/SubscriberController.php:4]: Please put the use statement below the copyright notice.

File: Classes/Domain/Model/Category.php [hunk 3]
```diff
@@ -1,6 +1,8 @@
 <?php
 namespace Slub\SlubEvents\Domain\Model;
 
+use TYPO3\CMS\Extbase\DomainObject\AbstractEntity;
```
Comment (albig) [inline Classes/Domain/Model/Category.php:4]: Please put the use statement below the copyright notice.

File: Classes/Domain/Model/Contact.php [hunk 4]
```diff
@@ -1,6 +1,8 @@
 <?php
 namespace Slub\SlubEvents\Domain\Model;
 
+use TYPO3\CMS\Extbase\DomainObject\AbstractEntity;
```
Comment (albig) [inline Classes/Domain/Model/Contact.php:4]: Please put the use statement below the copyright notice.

File: Classes/Domain/Model/Discipline.php [hunk 5]
```diff
@@ -1,6 +1,8 @@
 <?php
 namespace Slub\SlubEvents\Domain\Model;
 
+use TYPO3\CMS\Extbase\DomainObject\AbstractEntity;
```
Comment (albig) [inline Classes/Domain/Model/Discipline.php:4]: Please put the use statement below the copyright notice.

File: Classes/Domain/Model/Event.php [hunk 6]
```diff
@@ -1,6 +1,8 @@
 <?php
 namespace Slub\SlubEvents\Domain\Model;
 
+use TYPO3\CMS\Extbase\DomainObject\AbstractEntity;
```
Comment (albig) [inline Classes/Domain/Model/Event.php:4]: Please put the use statement below the copyright notice.

File: Classes/Domain/Repository/CategoryRepository.php [hunk 7]
```diff
@@ -1,6 +1,8 @@
 <?php
 namespace Slub\SlubEvents\Domain\Repository;
 
+use TYPO3\CMS\Extbase\Persistence\Repository;
```
Comment (albig) [inline Classes/Domain/Repository/CategoryRepository.php:4]: Please put the use statement below the copyright notice.

File: Classes/Domain/Repository/ContactRepository.php [hunk 8]
```diff
@@ -1,6 +1,8 @@
 <?php
 namespace Slub\SlubEvents\Domain\Repository;
 
+use TYPO3\CMS\Extbase\Persistence\Repository;
```
Comment (albig) [inline Classes/Domain/Repository/ContactRepository.php:4]: Please put the use statement below the copyright notice.

File: Classes/Domain/Repository/DisciplineRepository.php [hunk 9]
```diff
@@ -1,6 +1,8 @@
 <?php
 namespace Slub\SlubEvents\Domain\Repository;
 
+use TYPO3\CMS\Extbase\Persistence\Repository;
```
Comment (albig) [inline Classes/Domain/Repository/DisciplineRepository.php:4]: Please put the use statement below the copyright notice.

File: Classes/Domain/Repository/EventRepository.php [hunk 10]
```diff
@@ -1,6 +1,7 @@
 <?php
 namespace Slub\SlubEvents\Domain\Repository;
 
+    use TYPO3\CMS\Extbase\Persistence\Repository;
```
Comment (albig) [inline Classes/Domain/Repository/EventRepository.php:4]: Please put the use statement below the copyright notice.

File: Classes/Domain/Repository/LocationRepository.php [hunk 11]
```diff
@@ -1,6 +1,7 @@
 <?php
 namespace Slub\SlubEvents\Domain\Repository;
 
+use TYPO3\CMS\Extbase\Persistence\Repository;
```
Comment (albig) [inline Classes/Domain/Repository/LocationRepository.php:4]: Please put the use statement below the copyright notice.

File: Classes/Domain/Repository/SubscriberRepository.php [hunk 12]
```diff
@@ -1,6 +1,8 @@
 <?php
 namespace Slub\SlubEvents\Domain\Repository;
 
+use TYPO3\CMS\Extbase\Persistence\Repository;
```
Comment (albig) [inline Classes/Domain/Repository/SubscriberRepository.php:4]: Please put the use statement below the copyright notice.

File: Classes/Domain/Validator/SubscriberValidator.php [hunk 13]
```diff
@@ -1,6 +1,8 @@
 <?php
 namespace Slub\SlubEvents\Domain\Validator;
 
+use TYPO3\CMS\Extbase\Validation\Validator\AbstractValidator;
```
Comment (albig) [inline Classes/Domain/Validator/SubscriberValidator.php:4]: Please put the use statement below the copyright notice.

File: Classes/Helper/Form/Element/RecurringEventsElement.php [hunk 14]
```diff
@@ -1,5 +1,10 @@
 <?php
 namespace Slub\SlubEvents\Helper\Form\Element;
+use TYPO3\CMS\Core\Utility\GeneralUtility;
```
Comment (albig) [inline Classes/Helper/Form/Element/RecurringEventsElement.php:3]: Please put the use statement below the copyright notice.

File: Classes/Slots/HookPostProcessing.php [hunk 15]
```diff
@@ -1,5 +1,10 @@
 <?php
 namespace Slub\SlubEvents\Slots;
+use TYPO3\CMS\Extbase\Object\ObjectManager;
```
Comment (albig) [inline Classes/Slots/HookPostProcessing.php:3]: Please put the use statement below the copyright notice.

File: Classes/Task/CheckeventsTask.php [hunk 16]
```diff
@@ -1,5 +1,8 @@
 <?php
 namespace Slub\SlubEvents\Task;
+use TYPO3\CMS\Scheduler\Task\AbstractTask;
```
Comment (albig) [inline Classes/Task/CheckeventsTask.php:3]: Please put the use statement below the copyright notice.

File: Classes/Task/CheckeventsTaskAdditionalFieldProvider.php [hunk 17]
```diff
@@ -1,5 +1,7 @@
 <?php
 namespace Slub\SlubEvents\Task;
+use TYPO3\CMS\Scheduler\Controller\SchedulerModuleController;
```
Comment (albig) [inline Classes/Task/CheckeventsTaskAdditionalFieldProvider.php:3]: Please put the use statement below the copyright notice.

File: Classes/Task/CleanUpTask.php [hunk 18]
```diff
@@ -1,5 +1,11 @@
 <?php
 namespace Slub\SlubEvents\Task;
+use TYPO3\CMS\Scheduler\Task\AbstractTask;
```
Comment (albig) [inline Classes/Task/CleanUpTask.php:3]: Please put the use statement below the copyright notice.

File: Classes/Task/CleanUpTaskAdditionalFieldProvider.php [hunk 19]
```diff
@@ -1,5 +1,8 @@
 <?php
 namespace Slub\SlubEvents\Task;
+use TYPO3\CMS\Scheduler\Controller\SchedulerModuleController;
```
Comment (albig) [inline Classes/Task/CleanUpTaskAdditionalFieldProvider.php:3]: Please put the use statement below the copyright notice.

File: Classes/Task/StatisticsTask.php [hunk 20]
```diff
@@ -1,6 +1,9 @@
 <?php
 namespace Slub\SlubEvents\Task;
 
+use TYPO3\CMS\Scheduler\Task\AbstractTask;
```
Comment (albig) [inline Classes/Task/StatisticsTask.php:4]: Please put the use statement below the copyright notice.

File: Classes/ViewHelpers/Be/FunctionBarViewHelper.php [hunk 21]
```diff
@@ -1,5 +1,7 @@
 <?php
 namespace Slub\SlubEvents\ViewHelpers\Be;
+use TYPO3\CMS\Fluid\ViewHelpers\Be\AbstractBackendViewHelper;
```
Comment (albig) [inline Classes/ViewHelpers/Be/FunctionBarViewHelper.php:3]: Please put the use statement below the copyright notice.

File: Classes/ViewHelpers/Condition/GbEventsOfCategoryViewHelper.php [hunk 22]
```diff
@@ -1,6 +1,7 @@
 <?php
 namespace Slub\SlubEvents\ViewHelpers\Condition;
 
+use TYPO3Fluid\Fluid\Core\ViewHelper\AbstractViewHelper;
```
Comment (albig) [inline Classes/ViewHelpers/Condition/GbEventsOfCategoryViewHelper.php:4]: Please put the use statement below the copyright notice.

File: Classes/ViewHelpers/Condition/InArrayViewHelper.php [hunk 23]
```diff
@@ -1,6 +1,7 @@
 <?php
 namespace Slub\SlubEvents\ViewHelpers\Condition;
 
+use TYPO3Fluid\Fluid\Core\ViewHelper\AbstractViewHelper;
```
Comment (albig) [inline Classes/ViewHelpers/Condition/InArrayViewHelper.php:4]: Please put the use statement below the copyright notice.

File: Classes/ViewHelpers/Condition/IsPastEventViewHelper.php [hunk 24]
```diff
@@ -1,6 +1,7 @@
 <?php
 namespace Slub\SlubEvents\ViewHelpers\Condition;
 
+use TYPO3Fluid\Fluid\Core\ViewHelper\AbstractViewHelper;
```
Comment (albig) [inline Classes/ViewHelpers/Condition/IsPastEventViewHelper.php:4]: Please put the use statement below the copyright notice.

File: Classes/ViewHelpers/Condition/IsSubscriptionAllowedViewHelper.php [hunk 25]
```diff
@@ -1,6 +1,7 @@
 <?php
 namespace Slub\SlubEvents\ViewHelpers\Condition;
 
+use TYPO3Fluid\Fluid\Core\ViewHelper\AbstractViewHelper;
```
Comment (albig) [inline Classes/ViewHelpers/Condition/IsSubscriptionAllowedViewHelper.php:4]: Please put the use statement below the copyright notice.

File: Classes/ViewHelpers/Format/BookedPlacesViewHelper.php [hunk 26]
```diff
@@ -1,6 +1,7 @@
 <?php
 namespace Slub\SlubEvents\ViewHelpers\Format;
 
+use TYPO3Fluid\Fluid\Core\ViewHelper\AbstractViewHelper;
```
Comment (albig) [inline Classes/ViewHelpers/Format/BookedPlacesViewHelper.php:4]: Please put the use statement below the copyright notice.

File: Classes/ViewHelpers/Format/CsvEscapeViewHelper.php [hunk 27]
```diff
@@ -1,6 +1,7 @@
 <?php
 namespace Slub\SlubEvents\ViewHelpers\Format;
 
+use TYPO3Fluid\Fluid\Core\ViewHelper\AbstractViewHelper;
```
Comment (albig) [inline Classes/ViewHelpers/Format/CsvEscapeViewHelper.php:4]: Please put the use statement below the copyright notice.

File: Classes/ViewHelpers/Format/DateDiffViewHelper.php [hunk 28]
```diff
@@ -1,6 +1,7 @@
 <?php
 namespace Slub\SlubEvents\ViewHelpers\Format;
 
+use TYPO3Fluid\Fluid\Core\ViewHelper\AbstractViewHelper;
```
Comment (albig) [inline Classes/ViewHelpers/Format/DateDiffViewHelper.php:4]: Please put the use statement below the copyright notice.

File: Classes/ViewHelpers/Format/EditCodeViewHelper.php [hunk 29]
```diff
@@ -1,6 +1,7 @@
 <?php
 namespace Slub\SlubEvents\ViewHelpers\Format;
 
+use TYPO3Fluid\Fluid\Core\ViewHelper\AbstractViewHelper;
```
Comment (albig) [inline Classes/ViewHelpers/Format/EditCodeViewHelper.php:4]: Please put the use statement below the copyright notice.

File: Classes/ViewHelpers/Format/EventsOfCategoryViewHelper.php [hunk 30]
```diff
@@ -1,6 +1,7 @@
 <?php
 namespace Slub\SlubEvents\ViewHelpers\Format;
 
+use TYPO3Fluid\Fluid\Core\ViewHelper\AbstractViewHelper;
```
Comment (albig) [inline Classes/ViewHelpers/Format/EventsOfCategoryViewHelper.php:4]: Please put the use statement below the copyright notice.

File: Classes/ViewHelpers/Format/EventsOfDisciplineViewHelper.php [hunk 31]
```diff
@@ -1,6 +1,7 @@
 <?php
 namespace Slub\SlubEvents\ViewHelpers\Format;
 
+use TYPO3Fluid\Fluid\Core\ViewHelper\AbstractViewHelper;
```
Comment (albig) [inline Classes/ViewHelpers/Format/EventsOfDisciplineViewHelper.php:4]: Please put the use statement below the copyright notice.

File: Classes/ViewHelpers/Format/FreePlacesLeftViewHelper.php [hunk 32]
```diff
@@ -1,6 +1,7 @@
 <?php
 namespace Slub\SlubEvents\ViewHelpers\Format;
 
+use TYPO3Fluid\Fluid\Core\ViewHelper\AbstractViewHelper;
```
Comment (albig) [inline Classes/ViewHelpers/Format/FreePlacesLeftViewHelper.php:4]: Please put the use statement below the copyright notice.

File: Classes/ViewHelpers/Format/Fullcalendar/JsFooterViewHelper.php [hunk 33]
```diff
@@ -1,6 +1,7 @@
 <?php
 namespace Slub\SlubEvents\ViewHelpers\Format\Fullcalendar;
 
+use TYPO3Fluid\Fluid\Core\ViewHelper\AbstractViewHelper;
```
Comment (albig) [inline Classes/ViewHelpers/Format/Fullcalendar/JsFooterViewHelper.php:4]: Please put the use statement below the copyright notice.

File: Classes/ViewHelpers/Format/NewMonthTitleViewHelper.php [hunk 34]
```diff
@@ -1,6 +1,8 @@
 <?php
 namespace Slub\SlubEvents\ViewHelpers\Format;
 
+use TYPO3Fluid\Fluid\Core\ViewHelper\AbstractViewHelper;
```
Comment (albig) [inline Classes/ViewHelpers/Format/NewMonthTitleViewHelper.php:4]: Please put the use statement below the copyright notice.

File: Classes/ViewHelpers/Format/OneLineViewHelper.php [hunk 35]
```diff
@@ -1,6 +1,7 @@
 <?php
 namespace Slub\SlubEvents\ViewHelpers\Format;
 
+use TYPO3Fluid\Fluid\Core\ViewHelper\AbstractViewHelper;
```
Comment (albig) [inline Classes/ViewHelpers/Format/OneLineViewHelper.php:4]: Please put the use statement below the copyright notice.

File: Classes/ViewHelpers/Format/TrimViewHelper.php [hunk 36]
```diff
@@ -1,6 +1,7 @@
 <?php
 namespace Slub\SlubEvents\ViewHelpers\Format;
 
+use TYPO3Fluid\Fluid\Core\ViewHelper\AbstractViewHelper;
```
Comment (albig) [inline Classes/ViewHelpers/Format/TrimViewHelper.php:4]: Please put the use statement below the copyright notice.

File: Classes/ViewHelpers/Link/Is3dModelViewHelper.php [hunk 37]
```diff
@@ -1,6 +1,7 @@
 <?php
 namespace Slub\SlubEvents\ViewHelpers\Link;
 
+use TYPO3Fluid\Fluid\Core\ViewHelper\AbstractViewHelper;
```
Comment (albig) [inline Classes/ViewHelpers/Link/Is3dModelViewHelper.php:4]: Please put the use statement below the copyright notice.

File: Classes/ViewHelpers/Link/IslocalViewHelper.php [hunk 38]
```diff
@@ -1,6 +1,8 @@
 <?php
 namespace Slub\SlubEvents\ViewHelpers\Link;
 
+use TYPO3Fluid\Fluid\Core\ViewHelper\AbstractViewHelper;
```
Comment (albig) [inline Classes/ViewHelpers/Link/IslocalViewHelper.php:4]: Please put the use statement below the copyright notice.

**Review by albig — CHANGES_REQUESTED** [review]
Thank you for your work and sorry for the late reply.

Everything seems to work as expected.

Just a wish to make the formating constistently. I was used with TYPO3 9 to have the namespace above and the use statements below the copyright notice. I've seen, this has changed with TYPO3 10. Everything is now _below_ the copyright notice. This seems to be kept in TYPO3 11.

So, I would suggest, to make it consistently like TYPO3 10+.

I'm a little confused about Helper and Utility directory. Do we really need both? The "helper" are some helping utilities ;-) But that's not so important. I think, I understand your distinction.

**Comment by albig** [comment]
I don't know, if this makes a difference with rector. But we need no support of TYPO3 7.6 in this pull request anymore.

**Comment by PeaceSfiSh** [comment]
The ruleset definition from 7 to 9 is like a transition from v6-compatibility to v9-compatibility.
e.g. Typo3SetList::TYPO3_76 includes the rules for the migration from 6 LTS to 7 LTS
all three rulesets end in a better v9 LTS compatibility by removing old TCA-options or migrating them

**Pull request merged**